doc/reference/trd-uart.md

```diff
@@ -0,0 +1,540 @@
+`rval` of `Err(CANCEL)`.
+
+Reception early termination is necessary for UART virtualization. For
+example, suppose there are two UART clients. The first issues a read
+of 80 bytes.  After 20 bytes have been read, the second client issues
+a read of 40 bytes.  At this point, the virtualizer has to reduce the
+length of its outstanding read, from 60 (80-20) to 40 bytes. It needs
+to copy the 20 bytes read into the first client's buffer, the next 40
+bytes into both of their buffers, and the last 20 bytes read into the
+first client's buffer. It accomplishes this by calling `receive_abort`
+to terminate the 100-byte read, copying the bytes read from the
+resulting callback, then issuing a `receive_buffer` of 40 bytes.
+
+The valid return values for `receive_abort` are:
+  - `Callback(true)`: there was a reception outstanding and
+     it has been cancelled.  A callback with `Err(CANCEL)` will be called.
+  - `Callback(false)`: there was a reception outstanding but it
+     was not cancelled.  A callback will be called with an `rval` other than
+     `Err(CANCEL)`.
+  - `NoCallback`: there was no reception outstanding and the
+    implementation will not issue a callback.
+
+If there is no outstanding call to `receive_buffer` or
+`receive_character`, `receive_abort` MUST return `NoCallback`.
+
+4.2 `receive_character` and `received_character`
+===============================
+
+The `receive_character` method and `received_character` callback allow
+a client to perform character operations without buffer
+management. They receive a single UART character, where the character width
+is defined by the UART configuration and can be wider than 8
+bits.
+
+A UART implementation MAY choose to not implement `receive_character` and
+`received_character`.  There is a default implementation of `received_character`
+so clients that do not use `receive_character` do not have to implement a
+callback.
+
+If the UART returns `Ok(())` to a call to `receive_character`, it MUST make
+a `received_character` callback in the future, when it receives a character
+or some error occurs. Valid `Err` values of `receive_character` are:
+  - `BUSY`: the UART is busy with an outstanding call to
+    `receive_buffer` or `receive_character`.
+  - `OFF`: the UART is powered down or in a configuration that does
+    not allow UART reception (e.g., it is a USART in SPI mode).
+  - `NOSUPPORT`: `receive_character` operations are not supported.
+  - `FAIL`: some other error.
+
+5 Composite Traits
+===============================
+
+In addition to the 6 basic traits, the UART HIL defines several traits
+that use these basic traits as supertraits. These composite traits allow
+structures to refer to multiple pieces of UART functionality with a
+single reference and ensure that their implementations are coupled.
+
+```rust
+pub trait Uart<'a>: Configure + Configuration + Transmit<'a> + Receive<'a> {}
+pub trait UartData<'a>: Configuration + Transmit<'a> + Receive<'a> {}
```

Comment:
```suggestion
pub trait UartData<'a>:  Transmit<'a> + Receive<'a> {}
```
I thought the point of `UartData` was this it did not require `Configuration`?

Reply:
Good catch. Fixed.